### PR TITLE
add `make_lipschitz_mul_float`

### PR DESCRIPTION
**Pull request by Shoeboxam**
See #470

* Adds a `make_lipschitz_mul_float` constructor for multiplication of aggregated float values by a `constant`.
    * map: `d_out = d_in * constant + ulp(greatest_output)`
    * to be used to implement mean and variance

### CODE REVIEW

File: rust/src/trans/lipschitz_cast/mod.rs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    [u8, u16, u32, u64, u128, i8, i16, i32, i64, i128],
+    [f32, f64],
+    impl_greatest_cast_inf
+}
+// float to float
+impl GreatestDifference<f64> for f32 {
+    const C: Self = f32::INFINITY;
+}
+impl GreatestDifference<f32> for f64 {
+    const C: Self = f64::INFINITY;
```
Comment (andrewvyrros) [inline rust/src/trans/lipschitz_cast/mod.rs:148]: Why is this infinity, isn't f32 -> f64 an exact conversion?

**Review by andrewvyrros — COMMENTED** [review]
This looks like a helpful factoring. I can't really grok the details of `make_lipschitz_cast` and `GreatestDifference`, will need you to walk me through it.

File: rust/src/trans/lipschitz_cast/mod.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    ($tyi:ty, $tyo:ty) => {
+        impl GreatestDifference<$tyi> for $tyo {
+            const C: Self = Self::INFINITY;
+        }
+    };
+}
+
+use crate::traits::cartesian;
+// the three macros are applied to the upper triangle, diagonal, and lower triangle
+// conversion from any int type to any int type not on the diagonal has max distance 1
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/trans/lipschitz_cast/mod.rs:125]: I'm just going to split lipschitz_cast into it's own PR. I don't need it until the discrete laplace anyways.

File: python/test/test_trans.py [hunk 3]
```diff
@@ -291,3 +291,12 @@ def test_indexing():
...
     assert (binner >> indexer)(INT_DATA) == ['A', 'B', 'C', 'C', 'NA', 'NA', 'NA', 'NA', 'NA']
+
+
+def test_lipschitz_mul_float():
+    from opendp.trans import make_lipschitz_mul_float, make_sized_bounded_float_ordered_sum
+    trans = make_sized_bounded_float_ordered_sum(10, (0., 10.)) >> make_lipschitz_mul_float(1 / 10, (-3., 4.))
+
+    print(trans([3.] * 10))
+    print(trans.map(2))
+test_lipschitz_mul_float()
```
Comment (andrewvyrros) [inline python/test/test_trans.py:302]: Hard-wired test.

File: rust/src/trans/bootstrap.json [hunk 4]
```diff
@@ -667,6 +667,54 @@
...
+                    "args": ["T", "T"]
+                },
+                "hint": "Tuple[Any, Any]",
+                "description": "Tuple of inclusive lower and upper bounds of the input data."
+            },
+            {
+                "name": "D",
+                "default": "AllDomain<T>",
+                "generics": ["T"],
+                "description": "Domain of the identity function. Must be AllDomain<T> or VectorDomain<AllDomain<T>>",
```
Comment (andrewvyrros) [inline rust/src/trans/bootstrap.json:694]: Copy/pasted description from identity?

File: rust/src/trans/lipschitz_mul/mod.rs [hunk 5]
```diff
@@ -1,70 +1,122 @@
...
+    },
 };
 
-pub fn make_lipschitz_mul<D, M>(l: D::Atom) -> Fallible<Transformation<D, D, M, M>>
+use super::Float;
+
+#[cfg(feature = "ffi")]
+mod ffi;
+
+pub fn make_lipschitz_mul_float<D, M>(
```
Comment (andrewvyrros) [inline rust/src/trans/lipschitz_mul/mod.rs:19]: Should bounded be in the name somewhere now? Following the pattern of the sum constructors, `make_bounded_float_lipschitz_mul` or `make_bounded_lipschitz_float_mul`?

File: rust/src/trans/lipschitz_mul/mod.rs [hunk 6]
```diff
@@ -1,70 +1,122 @@
...
+    constant: D::Atom,
```
Comment (andrewvyrros) [inline rust/src/trans/lipschitz_mul/mod.rs:20]: Changing this to `constant` helps with readability.

**Review by andrewvyrros — COMMENTED** [review]
Removing the cast stuff seems good for now.

File: rust/src/trans/lipschitz_mul/mod.rs (hunk 5 above)
Comment (Shoeboxam) [reply rust/src/trans/lipschitz_mul/mod.rs:19]: In this case, the input domain is not bounded. I thought we used that naming to indicate something about the input domain? Technically, we could split this constructor into a lipschitz_clamp and a lipschitz_mul... But I don't see much utility in a lipschitz_clamp on its own.

Comment (andrewvyrros) [reply rust/src/trans/lipschitz_mul/mod.rs]: Ah yes thanks! But sounds like we want to move the position of `mul` to match the sum constructors.

**Review by andrewvyrros — APPROVED** [review]
This will be a nice building block to have.

**Pull request merged**